# Codebasics Resume Challange Project

## Tamil Nadu's 2026 Election Analysis

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

### Importing Datasets

In [142]:
constituency_master = pd.read_csv(r"D:\Codebasics Resume Challange\input_files_for_participants_rpc\data\constituency_master.csv")
turnout_2026 = pd.read_excel(r"D:\Codebasics Resume Challange\input_files_for_participants_rpc\data\Turnout_2026.xlsx")
tn_2021_results = pd.read_csv(r"D:\Codebasics Resume Challange\input_files_for_participants_rpc\data\tn_2021_results.csv")
tn_2026_results = pd.read_csv(r"D:\Codebasics Resume Challange\input_files_for_participants_rpc\data\tn_2026_results.csv")

### EDA 

In [28]:
tn_2021_results.shape

(4232, 8)

In [30]:
tn_2026_results.shape

(4257, 8)

#### From the above, we can say that there were 25 more candidates contested in 2026 
#### “The 2026 dataset contains a higher number of contesting candidates compared to 2021.”

## Question 1. The Geographic Story :- How does seat distribution shift across Tamil Nadu's six regions between 2021 & 2026 ? Where did each major formation gain or lose ground

In [79]:
# Firstly, we will create a new dataframe of winners of 2021 and 2026. This will create the data from candidate level to constituency level

In [172]:
winner_2021 = tn_2021_results.loc[
    tn_2021_results.groupby("constituency")["votes"].idxmax()
]

In [173]:
winner_2021

,constituency,ac_number,candidate,party,votes,turnout,reserved,region
610,Alandur,28,T.M.Anbarasan,DMK,116785,60.82,GEN,Chennai Metro
3264,Alangudi,182,SIVA.V.MEYYANATHAN,DMK,87935,78.47,GEN,Delta
4053,Alangulam,223,PAUL MANOJ PANDIAN,AIADMK,74153,77.40,GEN,South
4079,Ambasamudram,225,E.SUBAYA,AIADMK,85211,72.03,GEN,South
108,Ambattur,8,JOSEPH SAMUEL,DMK,114554,62.01,GEN,Chennai Metro
...,...,...,...,...,...,...,...,...
3196,Viralimalai,179,VIJAYA BASKER C,AIADMK,102179,85.43,GEN,Delta
3712,Virudhunagar,206,SEENIVASAN A.R.R,DMK,73297,71.27,GEN,South
461,Virugampakkam,22,AMV.PRABHAKARA RAJA,DMK,74351,57.59,GEN,Chennai Metro
2828,Vriddhachalam,152,Radhakrishnan,INC,77064,77.02,GEN,Central


In [144]:
winner_2026 = tn_2026_results.loc[
    tn_2026_results.groupby("constituency")["votes"].idxmax()
]

In [62]:
winner_2026

,constituency,ac_number,candidate,party,votes,turnout,reserved,region
633,Alandur,28,M.HARISH,TVK,112205,NaN,GEN,Chennai Metro
3347,Alangudi,182,SIVA.V.MEYYANATHAN,DMK,64929,NaN,GEN,Delta
4057,Alangulam,223,PAUL MANOJ PANDIAN,DMK,69170,NaN,GEN,South
4092,Ambasamudram,225,DR.ESAKKI SUBAYA,AIADMK,65589,NaN,GEN,South
114,Ambattur,8,BALAMURUGAN.G,TVK,133339,NaN,GEN,Chennai Metro
...,...,...,...,...,...,...,...,...
3274,Viralimalai,179,VIJAYABASKAR. C,AIADMK,105773,NaN,GEN,Delta
3737,Virudhunagar,206,SELVAM P,TVK,63653,NaN,GEN,South
489,Virugampakkam,22,SABARINATHAN.R,TVK,76092,NaN,GEN,Chennai Metro
2924,Vriddhachalam,152,PREMALLATHA VIJAYAKANT,DMDK,69351,NaN,GEN,Central


In [80]:
# Now, we will derieve new dataframe of region-wise party's seats for both year of 2021 & 2026

In [145]:
regional_2021 = winner_2021.groupby(
    ["region", "party"]
).size().reset_index(name="seats")

In [64]:
regional_2021

,region,party,seats
0,Central,AIADMK,15
1,Central,DMK,21
2,Central,INC,1
3,Central,PMK,3
4,Central,VCK,1
5,Chennai Metro,DMK,29
6,Chennai Metro,INC,3
7,Delta,AIADMK,4
8,Delta,CPI,1
9,Delta,CPI(M),2


In [146]:
regional_2026 = winner_2026.groupby(
    ["region","party"]
).size().reset_index(name='seats')

In [66]:
regional_2026

,region,party,seats
0,Central,AIADMK,15
1,Central,DMDK,1
2,Central,DMK,8
3,Central,PMK,3
4,Central,TVK,12
5,Central,VCK,2
6,Chennai Metro,AIADMK,1
7,Chennai Metro,DMK,2
8,Chennai Metro,TVK,29
9,Delta,AIADMK,4


In [81]:
# A geo_comapre dataframe which will tell us about the comparison of party's seat in 2021 & 2026. And will also show the respective change.

In [147]:
geo_compare = regional_2021.merge(
    regional_2026,
    on=["region", "party"],
    how="outer",
    suffixes=("_2021", "_2026")
)

In [148]:
geo_compare = geo_compare.fillna(0)


In [150]:
geo_compare["seat_change"]=(geo_compare["seats_2026"]-geo_compare["seats_2021"]) 

In [151]:
geo_compare

,region,party,seats_2021,seats_2026,seat_change
0,Central,AIADMK,15.0,15.0,0.0
1,Central,DMDK,0.0,1.0,1.0
2,Central,DMK,21.0,8.0,-13.0
3,Central,INC,1.0,0.0,-1.0
4,Central,PMK,3.0,3.0,0.0
5,Central,TVK,0.0,12.0,12.0
6,Central,VCK,1.0,2.0,1.0
7,Chennai Metro,AIADMK,0.0,1.0,1.0
8,Chennai Metro,DMK,29.0,2.0,-27.0
9,Chennai Metro,INC,3.0,0.0,-3.0


#### The most changing region was :-


In [152]:
geo_compare.sort_values(
    "seat_change",
    ascending=False
)

,region,party,seats_2021,seats_2026,seat_change
10,Chennai Metro,TVK,0.0,29.0,29.0
38,South,TVK,0.0,25.0,25.0
24,Kongu,TVK,0.0,16.0,16.0
31,North,TVK,0.0,15.0,15.0
5,Central,TVK,0.0,12.0,12.0
18,Delta,TVK,0.0,10.0,10.0
25,North,AIADMK,12.0,15.0,3.0
17,Delta,IUML,0.0,1.0,1.0
29,North,IUML,0.0,1.0,1.0
35,South,CPI(M),0.0,1.0,1.0


In [153]:
geo_compare.to_csv("geo_compare.csv", index=False)

### Mentioning the reserved seats for SC & ST

In [188]:
reserved_fix = {

    # SC Reserved Constituencies
    "Ponneri": "Ponneri (SC)",
    "Poonamallee": "Poonamallee (SC)",
    "Thiru-Vi-Ka-Nagar": "Thiru-Vi-Ka-Nagar(SC)",
    "Egmore": "Egmore (SC)",
    "Sriperumbudur": "Sriperumbudur (SC)",
    "Cheyyur": "Cheyyur (SC)",
    "Maduranthakam": "Madurantakam (SC)",
    "Arakonam": "Arakkonam (SC)",
    "Kilvaithinankuppam": "Kilvaithinankuppam(SC",
    "Gudiyatham": "Gudiyattam (SC)",
    "Uthangarai": "Uthangarai (SC)",
    "Harur": "Harur (SC)",
    "Chengam": "Chengam (SC)",
    "Vandavasi": "Vandavasi (SC)",
    "Tindivanam": "Tindivanam (SC)",
    "Vanur": "Vanur (SC)",
    "Kallakurichi": "Kallakurichi (SC)",
    "Gangavalli": "Gangavalli (SC)",
    "Attur": "Attur(SC)",
    "Rasipuram": "Rasipuram (SC)",
    "Dharapuram": "Dharapuram (SC)",
    "Bhavanisagar": "Bhavanisagar",
    "Gudalur": "Gudalur (SC)",
    "Avanashi": "Avanashi (SC)",
    "Valparai": "Valparai (SC)",
    "Nilakottai": "Nilakkottai (SC)",
    "Krishnarayapuram": "Krishnarayapuram(SC)",
    "Thuraiyur": "Thuraiyur (SC)",
    "Perambalur": "Perambalur (SC)",
    "Tittakudi": "Tittakudi (SC)",
    "Kattumannarkoil": "Kattumannarkoil(SC)",
    "Sirkali": "Sirkazhi (SC)",
    "Kilvelur": "Kilvelur (SC)",
    "Thiruthuraipundi": "Thiruthuraipoondi(SC)",
    "Thiruvidamarudur": "Thiruvidaimarudur",
    "Gandharvakottai": "Gandharvakottai(SC)",
    "Manamadurai": "Manamadurai(SC)",
    "Sholavandan": "Sholavandan(SC)",
    "Periyakulam": "Periyakulam",
    "Srivilliputhur": "Srivilliputhur(SC)",
    "Paramakudi": "Paramakudi (SC)",
    "Ottapidaram": "Ottapidaram(SC)",
    "Sankarankovil": "Sankarankovil(SC)",
    "Vasudevanallur": "Vasudevanallur(SC)",
    "Krishnarayapuram" : "Krishnarayapuram(SC)",

    # ST Reserved Constituencies
    "Yercaud": "Yercaud (ST)",
    "Sendamangalam": "Senthamangalam(ST)",

    #Misseplled 
    "Arantangi" : "Aranthangi",
    "Gummidipundi" : "Gummidipoondi",
    "Sholinghur" : "Sholingur",
    "Vaniayambadi" : "Vaniyambadi",
    "Palacode" : "Palacodu",
    "Rishivandiam" : "Rishivandiyam",
    "Ulundurpet" : "Ulundurpettai",
    "Villupuram" : "Viluppuram",
    "Edapadi" : "Edappadi",
    "Mettupalayam" : "Mettuppalayam",
    "Tiruchengode" : "Tiruchengodu",
    "Modakurichi" : "Modakkurichi",
    "Udumalpet" : "Udumalaipettai",
    "Tiruchirapalli (East)" : "Tiruchirapalli",
    "Tiruchirapalli (West)" : "Tiruchirapalli",
    "Thiruverambur" : "Thiruverumbur",
    "Orathanad" : "Orathanadu",
    "Aruppukotai" : "Aruppukkottai",
    "Mudukulathur" : "Mudhukulathur",
    "Palayamcottai" : "Palayamkottai",
    "Chepauk-Thiruvallikeni" : "Chepauk-Thiruvalliken",
    "Dr. Radhakrishnan Nagar" : "Dr.Radhakrishnan Naga",
    "Thiruvottiyur" : "Tiruvottiyur",
    "Tiruchirapalli" : "Tiruchirappalli",
    "Coimbatore South" : "Coimbatore(South)",
    "Theayagaraya Nagar" : "Thiyagarayanagar",
    "Coimbatore North" : "Coimbatore(North)"
    
}

In [189]:
winner_2021["constituency"] = winner_2021["constituency"].replace(reserved_fix)

In [190]:
winner_2021.to_csv("constituency_win_2021.csv", index=False)

In [192]:
winner_2026["constituency"]=winner_2026["constituency"].replace(reserved_fix)

In [194]:
winner_2026.to_csv("constituency_win_2026.csv", index=False)

In [195]:
winner_2021

,constituency,ac_number,candidate,party,votes,turnout,reserved,region
610,Alandur,28,T.M.Anbarasan,DMK,116785,60.82,GEN,Chennai Metro
3264,Alangudi,182,SIVA.V.MEYYANATHAN,DMK,87935,78.47,GEN,Delta
4053,Alangulam,223,PAUL MANOJ PANDIAN,AIADMK,74153,77.40,GEN,South
4079,Ambasamudram,225,E.SUBAYA,AIADMK,85211,72.03,GEN,South
108,Ambattur,8,JOSEPH SAMUEL,DMK,114554,62.01,GEN,Chennai Metro
...,...,...,...,...,...,...,...,...
3196,Viralimalai,179,VIJAYA BASKER C,AIADMK,102179,85.43,GEN,Delta
3712,Virudhunagar,206,SEENIVASAN A.R.R,DMK,73297,71.27,GEN,South
461,Virugampakkam,22,AMV.PRABHAKARA RAJA,DMK,74351,57.59,GEN,Chennai Metro
2828,Vriddhachalam,152,Radhakrishnan,INC,77064,77.02,GEN,Central


In [ ]:
geo_compare = regional_2021.merge(
    regional_2026,
    on=["region", "party"],
    how="outer",
    suffixes=("_2021", "_2026")
)

In [207]:
win_distribution = winner_2021[["ac_number","constituency", "party", "region"]].merge(
    winner_2026[["ac_number","constituency", "party", "region"]],
    on=["constituency"],
    how="outer",
    suffixes=("_2021","_2026")
)

In [243]:
win_distribution

,ac_number,constituency,party_2021,party_2026,region
0,28.0,Alandur,DMK,TVK,Chennai Metro
1,182.0,Alangudi,DMK,DMK,Delta
2,223.0,Alangulam,AIADMK,DMK,South
3,225.0,Ambasamudram,AIADMK,AIADMK,South
4,8.0,Ambattur,DMK,TVK,Chennai Metro
...,...,...,...,...,...
243,179.0,Viralimalai,AIADMK,AIADMK,Delta
244,206.0,Virudhunagar,DMK,TVK,South
245,22.0,Virugampakkam,DMK,TVK,Chennai Metro
246,152.0,Vriddhachalam,INC,DMDK,Central


In [450]:
redistribution_count = (
    win_distribution
    .groupby(
        ["party_2021", "party_2026", "region"]
    )
    .size()
    .reset_index(name="count")
    .sort_values("count", ascending=False)
)

In [453]:
#redistribution is saved as "party_flips_region"

redistribution_count.to_csv("party_flips_region.csv", index = False)

In [249]:
redistribution_count[
    (redistribution_count["party_2021"] == "DMK") &
    (redistribution_count["party_2026"] == "TVK")
]

,party_2021,party_2026,count
17,DMK,TVK,60


In [ ]:
##win_distribution table

In [263]:
win_distribution = pd.read_csv(r"D:\Codebasics Resume Challange\My_Work\win_comparison.csv")

In [446]:
redistribution_count = (
    win_distribution
    .groupby(
        ["party_2021", "party_2026"]
    )
    .size()
    .reset_index(name="count")
    .sort_values("count", ascending=False)
)

## redistribution_count and party_flips, both table are same. But redistribution_count(party_flips_region) has region column added

### The Flip Story

#### Question :- In how many constituencies the winning party changed between 2021 & 2026. Visaulize the shifts and what pattern emerges.

INSIGHTS 
- Winner changes were concentrated in specific region
- Certain party transitions occur significantly more frequently than others
- Some regions seems to be stable inspite of having high electoral volatility in other regions

In [264]:
party_flips = (
    win_distribution
    .groupby(
        ["party_2021","party_2026"]
    )
    .size()
    .reset_index(name="count")
)

In [271]:
party_flips= redistribution_count[redistribution_count["party_2021"] != redistribution_count["party_2026"]]

In [272]:
party_flips.sort_values("count", ascending=False)

,party_2021,party_2026,count
17,DMK,TVK,65
4,AIADMK,TVK,26
11,DMK,AIADMK,22
1,AIADMK,DMK,15
22,INC,TVK,11
16,DMK,PMK,3
23,PMK,AIADMK,2
7,BJP,TVK,2
25,PMK,TVK,2
6,BJP,DMK,2


In [273]:
## The above thing will be used in sankey diagram visual in power bi and presentation.
party_flips.to_csv("party_flips.csv", index=False)

### Margin of Victory Story

### Question :- How did the average margin of victory change between 2021 and 2026? In how many constituencies did the winning candidate secure more than 50% of the valid votes polled? In how many did the winning candidate receive less than 35% of the valid votes polled?

In [339]:
#margin 2021

margin_2021=tn_2021_results.sort_values(["constituency", "votes"],ascending=[True,False])

In [340]:
margin_2021

,constituency,ac_number,candidate,party,votes,turnout,reserved,region
610,Alandur,28,T.M.Anbarasan,DMK,116785,60.82,GEN,Chennai Metro
611,Alandur,28,B.Valarmathi,AIADMK,76214,60.82,GEN,Chennai Metro
612,Alandur,28,Sarathbabu,MNM,21117,60.82,GEN,Chennai Metro
613,Alandur,28,Dr.R.Karthikeyan,NTK,16506,60.82,GEN,Chennai Metro
614,Alandur,28,None Of The Above,NOTA,1888,60.82,GEN,Chennai Metro
...,...,...,...,...,...,...,...,...
1491,Yercaud,83,T. Priya,IND,289,83.09,ST,Central
1492,Yercaud,83,Santhosam,IND,167,83.09,ST,Central
1493,Yercaud,83,S.T. Rajendhiran,IND,165,83.09,ST,Central
1494,Yercaud,83,P. SATHISHKUMAR,IND,148,83.09,ST,Central


In [341]:
# Top 2 candidates
top2_2021 = (margin_2021.groupby("constituency").head(2))

In [342]:
top2_2021

,constituency,ac_number,candidate,party,votes,turnout,reserved,region
610,Alandur,28,T.M.Anbarasan,DMK,116785,60.82,GEN,Chennai Metro
611,Alandur,28,B.Valarmathi,AIADMK,76214,60.82,GEN,Chennai Metro
3264,Alangudi,182,SIVA.V.MEYYANATHAN,DMK,87935,78.47,GEN,Delta
3265,Alangudi,182,DHARMA THANGAVEL,AIADMK,62088,78.47,GEN,Delta
4053,Alangulam,223,PAUL MANOJ PANDIAN,AIADMK,74153,77.40,GEN,South
...,...,...,...,...,...,...,...,...
462,Virugampakkam,22,VIRUGAI V.N.RAVI,AIADMK,55984,57.59,GEN,Chennai Metro
2828,Vriddhachalam,152,Radhakrishnan,INC,77064,77.02,GEN,Central
2829,Vriddhachalam,152,Karthikeyan. J,PMK,76202,77.02,GEN,Central
1482,Yercaud,83,G. Chitra,AIADMK,121561,83.09,ST,Central


In [ ]:
top2_2021["rank"] = (
    top2_2021
    .groupby("constituency")["votes"]
    .rank(method="first", ascending=False)
)

In [542]:
top2_2021[top2_2021['ac_number']==185]

,constituency,ac_number,candidate,party,votes,turnout,reserved,region,rank
3313,Tiruppattur,185,KR.PERIYAKARUPPAN,DMK,103682,71.98,GEN,South,1.0


In [345]:
## Creating winner and runner up table 
winner_2021 = top2_2021[top2_2021["rank"]==1][["ac_number", "constituency", "party","votes", "reserved", "region"]]

In [346]:
winner_2021

,ac_number,constituency,party,votes,reserved,region
610,28,Alandur,DMK,116785,GEN,Chennai Metro
3264,182,Alangudi,DMK,87935,GEN,Delta
4053,223,Alangulam,AIADMK,74153,GEN,South
4079,225,Ambasamudram,AIADMK,85211,GEN,South
108,8,Ambattur,DMK,114554,GEN,Chennai Metro
...,...,...,...,...,...,...
3196,179,Viralimalai,AIADMK,102179,GEN,Delta
3712,206,Virudhunagar,DMK,73297,GEN,South
461,22,Virugampakkam,DMK,74351,GEN,Chennai Metro
2828,152,Vriddhachalam,INC,77064,GEN,Central


In [347]:
runner_2021 = top2_2021[top2_2021["rank"]==2][["ac_number","constituency","party","votes","reserved","region"]]

In [348]:
runner_2021

,ac_number,constituency,party,votes,reserved,region
611,28,Alandur,AIADMK,76214,GEN,Chennai Metro
3265,182,Alangudi,AIADMK,62088,GEN,Delta
4054,223,Alangulam,DMK,70614,GEN,South
4080,225,Ambasamudram,DMK,68296,GEN,South
109,8,Ambattur,AIADMK,72408,GEN,Chennai Metro
...,...,...,...,...,...,...
3197,179,Viralimalai,DMK,78581,GEN,Delta
3713,206,Virudhunagar,BJP,51958,GEN,South
462,22,Virugampakkam,AIADMK,55984,GEN,Chennai Metro
2829,152,Vriddhachalam,PMK,76202,GEN,Central


In [349]:
#creating a margin_2021 table
margin_2021 = winner_2021.merge(runner_2021, on = "constituency", suffixes = ("_winner", "_runner"))

In [350]:
margin_2021

,ac_number_winner,constituency,party_winner,votes_winner,reserved_winner,region_winner,ac_number_runner,party_runner,votes_runner,reserved_runner,region_runner
0,28,Alandur,DMK,116785,GEN,Chennai Metro,28,AIADMK,76214,GEN,Chennai Metro
1,182,Alangudi,DMK,87935,GEN,Delta,182,AIADMK,62088,GEN,Delta
2,223,Alangulam,AIADMK,74153,GEN,South,223,DMK,70614,GEN,South
3,225,Ambasamudram,AIADMK,85211,GEN,South,225,DMK,68296,GEN,South
4,8,Ambattur,DMK,114554,GEN,Chennai Metro,8,AIADMK,72408,GEN,Chennai Metro
...,...,...,...,...,...,...,...,...,...,...,...
228,179,Viralimalai,AIADMK,102179,GEN,Delta,179,DMK,78581,GEN,Delta
229,206,Virudhunagar,DMK,73297,GEN,South,206,BJP,51958,GEN,South
230,22,Virugampakkam,DMK,74351,GEN,Chennai Metro,22,AIADMK,55984,GEN,Chennai Metro
231,152,Vriddhachalam,INC,77064,GEN,Central,152,PMK,76202,GEN,Central


In [351]:
margin_2021["margin_votes"] = margin_2021["votes_winner"]-margin_2021["votes_runner"]

In [408]:
total_votes_2021 = (tn_2021_results.groupby("constituency")["votes"].sum().reset_index())

In [409]:
total_votes_2021

,constituency,votes
0,Alandur,237751
1,Alangudi,171842
2,Alangulam,203478
3,Ambasamudram,177681
4,Ambattur,240299
...,...,...
228,Viralimalai,193423
229,Virudhunagar,161717
230,Virugampakkam,169087
231,Vriddhachalam,196734


In [361]:
#merging total votes with margin_2021 table
margin_2021=margin_2021.merge(total_votes, on= "constituency",suffixes=("_winner", "_runner"))

In [362]:
margin_2021

,ac_number_winner,constituency,party_winner,votes_winner,reserved_winner,region_winner,ac_number_runner,party_runner,votes_runner,reserved_runner,region_runner,margin_votes,votes_x,votes_y,votes
0,28,Alandur,DMK,116785,GEN,Chennai Metro,28,AIADMK,76214,GEN,Chennai Metro,40571,237751,237751,237751
1,182,Alangudi,DMK,87935,GEN,Delta,182,AIADMK,62088,GEN,Delta,25847,171842,171842,171842
2,223,Alangulam,AIADMK,74153,GEN,South,223,DMK,70614,GEN,South,3539,203478,203478,203478
3,225,Ambasamudram,AIADMK,85211,GEN,South,225,DMK,68296,GEN,South,16915,177681,177681,177681
4,8,Ambattur,DMK,114554,GEN,Chennai Metro,8,AIADMK,72408,GEN,Chennai Metro,42146,240299,240299,240299
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
228,179,Viralimalai,AIADMK,102179,GEN,Delta,179,DMK,78581,GEN,Delta,23598,193423,193423,193423
229,206,Virudhunagar,DMK,73297,GEN,South,206,BJP,51958,GEN,South,21339,161717,161717,161717
230,22,Virugampakkam,DMK,74351,GEN,Chennai Metro,22,AIADMK,55984,GEN,Chennai Metro,18367,169087,169087,169087
231,152,Vriddhachalam,INC,77064,GEN,Central,152,PMK,76202,GEN,Central,862,196734,196734,196734


In [366]:
margin_2021=margin_2021.drop(columns={"votes_x","votes_y", "region_winner", "ac_number_runner", "reserved_winner"})

In [374]:
margin_2021=margin_2021.rename(columns={"ac_number_winner" : "ac_number", "reserved_runner":"reserved", "region_runner" : "region", "votes" : "total_votes" })

In [375]:
margin_2021

,ac_number,constituency,party_winner,votes_winner,party_runner,votes_runner,reserved,region,margin_votes,total_votes
0,28,Alandur,DMK,116785,AIADMK,76214,GEN,Chennai Metro,40571,237751
1,182,Alangudi,DMK,87935,AIADMK,62088,GEN,Delta,25847,171842
2,223,Alangulam,AIADMK,74153,DMK,70614,GEN,South,3539,203478
3,225,Ambasamudram,AIADMK,85211,DMK,68296,GEN,South,16915,177681
4,8,Ambattur,DMK,114554,AIADMK,72408,GEN,Chennai Metro,42146,240299
...,...,...,...,...,...,...,...,...,...,...
228,179,Viralimalai,AIADMK,102179,DMK,78581,GEN,Delta,23598,193423
229,206,Virudhunagar,DMK,73297,BJP,51958,GEN,South,21339,161717
230,22,Virugampakkam,DMK,74351,AIADMK,55984,GEN,Chennai Metro,18367,169087
231,152,Vriddhachalam,INC,77064,PMK,76202,GEN,Central,862,196734


In [378]:
margin_2021["vote_margin%"]=(margin_2021["margin_votes"]/margin_2021["total_votes"])*100

In [379]:
margin_2021

,ac_number,constituency,party_winner,votes_winner,party_runner,votes_runner,reserved,region,margin_votes,total_votes,vote_margin%
0,28,Alandur,DMK,116785,AIADMK,76214,GEN,Chennai Metro,40571,237751,17.064492
1,182,Alangudi,DMK,87935,AIADMK,62088,GEN,Delta,25847,171842,15.041142
2,223,Alangulam,AIADMK,74153,DMK,70614,GEN,South,3539,203478,1.739254
3,225,Ambasamudram,AIADMK,85211,DMK,68296,GEN,South,16915,177681,9.519870
4,8,Ambattur,DMK,114554,AIADMK,72408,GEN,Chennai Metro,42146,240299,17.538983
...,...,...,...,...,...,...,...,...,...,...,...
228,179,Viralimalai,AIADMK,102179,DMK,78581,GEN,Delta,23598,193423,12.200204
229,206,Virudhunagar,DMK,73297,BJP,51958,GEN,South,21339,161717,13.195273
230,22,Virugampakkam,DMK,74351,AIADMK,55984,GEN,Chennai Metro,18367,169087,10.862455
231,152,Vriddhachalam,INC,77064,PMK,76202,GEN,Central,862,196734,0.438155


- from margin%, we know that the first candidate won was won by xx$ more as compared to the second placed candidate

In [380]:
margin_2021["winner_share%"]= (margin_2021["votes_winner"]/margin_2021["total_votes"])*100

In [381]:
margin_2021

,ac_number,constituency,party_winner,votes_winner,party_runner,votes_runner,reserved,region,margin_votes,total_votes,vote_margin%,winner_share%
0,28,Alandur,DMK,116785,AIADMK,76214,GEN,Chennai Metro,40571,237751,17.064492,49.120719
1,182,Alangudi,DMK,87935,AIADMK,62088,GEN,Delta,25847,171842,15.041142,51.172007
2,223,Alangulam,AIADMK,74153,DMK,70614,GEN,South,3539,203478,1.739254,36.442760
3,225,Ambasamudram,AIADMK,85211,DMK,68296,GEN,South,16915,177681,9.519870,47.957294
4,8,Ambattur,DMK,114554,AIADMK,72408,GEN,Chennai Metro,42146,240299,17.538983,47.671443
...,...,...,...,...,...,...,...,...,...,...,...,...
228,179,Viralimalai,AIADMK,102179,DMK,78581,GEN,Delta,23598,193423,12.200204,52.826706
229,206,Virudhunagar,DMK,73297,BJP,51958,GEN,South,21339,161717,13.195273,45.324239
230,22,Virugampakkam,DMK,74351,AIADMK,55984,GEN,Chennai Metro,18367,169087,10.862455,43.972038
231,152,Vriddhachalam,INC,77064,PMK,76202,GEN,Central,862,196734,0.438155,39.171673


In [ ]:
## winner_share% tell us about -- What percentage of all valid votes were won by the winning candidate.

### now, same for the 2026

In [387]:
#margin 2026

margin_2026=tn_2026_results.sort_values(["constituency", "votes"],ascending=[True,False])

In [388]:
margin_2026

,constituency,ac_number,candidate,party,votes,turnout,reserved,region
633,Alandur,28,M.HARISH,TVK,112205,NaN,GEN,Chennai Metro
634,Alandur,28,T.M.ANBARASAN,DMK,82596,NaN,GEN,Chennai Metro
635,Alandur,28,S.SARAVANAN,AIADMK,52243,NaN,GEN,Chennai Metro
636,Alandur,28,Y.G.MAHALAKSHMI,NTK,10405,NaN,GEN,Chennai Metro
637,Alandur,28,NOTA,NOTA,1263,NaN,GEN,Chennai Metro
...,...,...,...,...,...,...,...,...
1602,Yercaud (ST),83,THANGADURAI. N,IND,294,NaN,ST,Central
1603,Yercaud (ST),83,VENGATESHSAN. T,Tamizhaga Vaazhvurimai Katchi,235,NaN,ST,Central
1604,Yercaud (ST),83,VIJAYAKUMAR. N,IND,211,NaN,ST,Central
1605,Yercaud (ST),83,KALIYAPPAN. A,IND,123,NaN,ST,Central


In [389]:
# Top 2 candidates
top2_2026 = (margin_2026.groupby("constituency").head(2))

In [390]:
top2_2026

,constituency,ac_number,candidate,party,votes,turnout,reserved,region
633,Alandur,28,M.HARISH,TVK,112205,NaN,GEN,Chennai Metro
634,Alandur,28,T.M.ANBARASAN,DMK,82596,NaN,GEN,Chennai Metro
3347,Alangudi,182,SIVA.V.MEYYANATHAN,DMK,64929,NaN,GEN,Delta
3348,Alangudi,182,KANDASAMY,TVK,51952,NaN,GEN,Delta
4057,Alangulam,223,PAUL MANOJ PANDIAN,DMK,69170,NaN,GEN,South
...,...,...,...,...,...,...,...,...
490,Virugampakkam,22,PRABHAKARA RAJA.A.M.V,DMK,49006,NaN,GEN,Chennai Metro
2924,Vriddhachalam,152,PREMALLATHA VIJAYAKANT,DMDK,69351,NaN,GEN,Central
2925,Vriddhachalam,152,VIJAY.S,TVK,66964,NaN,GEN,Central
1593,Yercaud (ST),83,USHARANI. P,AIADMK,87772,NaN,ST,Central


In [ ]:
top2_2026["rank"] = (
    top2_2026
    .groupby("constituency")["votes"]
    .rank(method="first", ascending=False)
)

In [392]:
top2_2026

,constituency,ac_number,candidate,party,votes,turnout,reserved,region,rank
633,Alandur,28,M.HARISH,TVK,112205,NaN,GEN,Chennai Metro,1.0
634,Alandur,28,T.M.ANBARASAN,DMK,82596,NaN,GEN,Chennai Metro,2.0
3347,Alangudi,182,SIVA.V.MEYYANATHAN,DMK,64929,NaN,GEN,Delta,1.0
3348,Alangudi,182,KANDASAMY,TVK,51952,NaN,GEN,Delta,2.0
4057,Alangulam,223,PAUL MANOJ PANDIAN,DMK,69170,NaN,GEN,South,1.0
...,...,...,...,...,...,...,...,...,...
490,Virugampakkam,22,PRABHAKARA RAJA.A.M.V,DMK,49006,NaN,GEN,Chennai Metro,2.0
2924,Vriddhachalam,152,PREMALLATHA VIJAYAKANT,DMDK,69351,NaN,GEN,Central,1.0
2925,Vriddhachalam,152,VIJAY.S,TVK,66964,NaN,GEN,Central,2.0
1593,Yercaud (ST),83,USHARANI. P,AIADMK,87772,NaN,ST,Central,1.0


In [393]:
## Creating winner and runner up table 
winner_2026 = top2_2026[top2_2026["rank"]==1][["ac_number", "constituency", "party","votes", "reserved", "region"]]

In [394]:
winner_2026

,ac_number,constituency,party,votes,reserved,region
633,28,Alandur,TVK,112205,GEN,Chennai Metro
3347,182,Alangudi,DMK,64929,GEN,Delta
4057,223,Alangulam,DMK,69170,GEN,South
4092,225,Ambasamudram,AIADMK,65589,GEN,South
114,8,Ambattur,TVK,133339,GEN,Chennai Metro
...,...,...,...,...,...,...
3274,179,Viralimalai,AIADMK,105773,GEN,Delta
3737,206,Virudhunagar,TVK,63653,GEN,South
489,22,Virugampakkam,TVK,76092,GEN,Chennai Metro
2924,152,Vriddhachalam,DMDK,69351,GEN,Central


In [395]:
runner_2026 = top2_2026[top2_2026["rank"]==2][["ac_number","constituency","party","votes","reserved","region"]]

In [396]:
runner_2026

,ac_number,constituency,party,votes,reserved,region
634,28,Alandur,DMK,82596,GEN,Chennai Metro
3348,182,Alangudi,TVK,51952,GEN,Delta
4058,223,Alangulam,AIADMK,61372,GEN,South
4093,225,Ambasamudram,INC,55344,GEN,South
115,8,Ambattur,DMK,74558,GEN,Chennai Metro
...,...,...,...,...,...,...
3275,179,Viralimalai,TVK,43700,GEN,Delta
3738,206,Virudhunagar,DMDK,54262,GEN,South
490,22,Virugampakkam,DMK,49006,GEN,Chennai Metro
2925,152,Vriddhachalam,TVK,66964,GEN,Central


In [494]:
#creating a margin_2026 table
margin_2026 = winner_2026.merge(runner_2026, on = "ac_number", suffixes = ("_winner", "_runner"))

In [496]:
margin_2026

,ac_number,constituency_winner,party_winner,votes_winner,reserved_winner,region_winner,constituency_runner,party_runner,votes_runner,reserved_runner,region_runner
0,28,Alandur,TVK,112205,GEN,Chennai Metro,Alandur,DMK,82596,GEN,Chennai Metro
1,182,Alangudi,DMK,64929,GEN,Delta,Alangudi,TVK,51952,GEN,Delta
2,223,Alangulam,DMK,69170,GEN,South,Alangulam,AIADMK,61372,GEN,South
3,225,Ambasamudram,AIADMK,65589,GEN,South,Ambasamudram,INC,55344,GEN,South
4,8,Ambattur,TVK,133339,GEN,Chennai Metro,Ambattur,DMK,74558,GEN,Chennai Metro
...,...,...,...,...,...,...,...,...,...,...,...
227,179,Viralimalai,AIADMK,105773,GEN,Delta,Viralimalai,TVK,43700,GEN,Delta
228,206,Virudhunagar,TVK,63653,GEN,South,Virudhunagar,DMDK,54262,GEN,South
229,22,Virugampakkam,TVK,76092,GEN,Chennai Metro,Virugampakkam,DMK,49006,GEN,Chennai Metro
230,152,Vriddhachalam,DMDK,69351,GEN,Central,Vriddhachalam,TVK,66964,GEN,Central


In [497]:
margin_2026=margin_2026.drop(columns={"reserved_winner", "region_winner", "constituency_runner"})

In [498]:
margin_2026=margin_2026.rename(columns={"constituency_winner":"constituency", "reserved_runner":"reserved", "region_runner":"region"})

In [499]:
margin_2026

,ac_number,constituency,party_winner,votes_winner,party_runner,votes_runner,reserved,region
0,28,Alandur,TVK,112205,DMK,82596,GEN,Chennai Metro
1,182,Alangudi,DMK,64929,TVK,51952,GEN,Delta
2,223,Alangulam,DMK,69170,AIADMK,61372,GEN,South
3,225,Ambasamudram,AIADMK,65589,INC,55344,GEN,South
4,8,Ambattur,TVK,133339,DMK,74558,GEN,Chennai Metro
...,...,...,...,...,...,...,...,...
227,179,Viralimalai,AIADMK,105773,TVK,43700,GEN,Delta
228,206,Virudhunagar,TVK,63653,DMDK,54262,GEN,South
229,22,Virugampakkam,TVK,76092,DMK,49006,GEN,Chennai Metro
230,152,Vriddhachalam,DMDK,69351,TVK,66964,GEN,Central


In [500]:
margin_2026["margin_votes"]=margin_2026["votes_winner"]-margin_2026["votes_runner"]

In [501]:
margin_2026

,ac_number,constituency,party_winner,votes_winner,party_runner,votes_runner,reserved,region,margin_votes
0,28,Alandur,TVK,112205,DMK,82596,GEN,Chennai Metro,29609
1,182,Alangudi,DMK,64929,TVK,51952,GEN,Delta,12977
2,223,Alangulam,DMK,69170,AIADMK,61372,GEN,South,7798
3,225,Ambasamudram,AIADMK,65589,INC,55344,GEN,South,10245
4,8,Ambattur,TVK,133339,DMK,74558,GEN,Chennai Metro,58781
...,...,...,...,...,...,...,...,...,...
227,179,Viralimalai,AIADMK,105773,TVK,43700,GEN,Delta,62073
228,206,Virudhunagar,TVK,63653,DMDK,54262,GEN,South,9391
229,22,Virugampakkam,TVK,76092,DMK,49006,GEN,Chennai Metro,27086
230,152,Vriddhachalam,DMDK,69351,TVK,66964,GEN,Central,2387


In [502]:
total_votes_2026 = (tn_2026_results.groupby("ac_number")["votes"].sum().reset_index())

In [503]:
total_votes_2026

,ac_number,votes
0,1,232522
1,2,226819
2,3,238798
3,4,226428
4,5,308894
...,...,...
229,230,196280
230,231,199013
231,232,180555
232,233,174927


In [504]:
#merging total votes with margin_2021 table
margin_2026=margin_2026.merge(total_votes_2026, on= "ac_number")

In [508]:
margin_2026=margin_2026.rename(columns={"votes":"total_votes"})

In [509]:
margin_2026["margin%"]=(margin_2026["margin_votes"]/margin_2026["total_votes"])*100

In [510]:
margin_2026["winner_share%"]=(margin_2026["votes_winner"]/margin_2026["total_votes"])*100

In [511]:
margin_2026

,ac_number,constituency,party_winner,votes_winner,party_runner,votes_runner,reserved,region,margin_votes,total_votes,margin%,winner_share%
0,28,Alandur,TVK,112205,DMK,82596,GEN,Chennai Metro,29609,261570,11.319723,42.896739
1,182,Alangudi,DMK,64929,TVK,51952,GEN,Delta,12977,180483,7.190151,35.975133
2,223,Alangulam,DMK,69170,AIADMK,61372,GEN,South,7798,220857,3.530791,31.318908
3,225,Ambasamudram,AIADMK,65589,INC,55344,GEN,South,10245,191195,5.358404,34.304767
4,8,Ambattur,TVK,133339,DMK,74558,GEN,Chennai Metro,58781,259986,22.609294,51.286992
...,...,...,...,...,...,...,...,...,...,...,...,...
227,179,Viralimalai,AIADMK,105773,TVK,43700,GEN,Delta,62073,203921,30.439729,51.869597
228,206,Virudhunagar,TVK,63653,DMDK,54262,GEN,South,9391,167832,5.595476,37.926617
229,22,Virugampakkam,TVK,76092,DMK,49006,GEN,Chennai Metro,27086,172831,15.671957,44.026824
230,152,Vriddhachalam,DMDK,69351,TVK,66964,GEN,Central,2387,209186,1.141090,33.152792


In [583]:
margin_2026 = pd.read_csv(r"D:\Codebasics Resume Challange\My_Work\margin_2026.csv")
margin_2021 = pd.read_csv(r"D:\Codebasics Resume Challange\My_Work\margin_2021.csv")

## The new margin table after validating the data from tn_results dataset of codebasics.

In [584]:
## Average vote margin of 2021 and 2026

In [585]:
margin_2026["margin%"].mean()

np.float64(7.6723245623034195)

In [586]:
margin_2021["vote_margin%"].mean()

np.float64(11.696277672542735)

In [587]:
## winners with less competitiveness for both 2021 & 2026

In [588]:
(margin_2021["winner_share%"] > 50).sum()

np.int64(70)

In [589]:
(margin_2026["winner_share%"] > 50).sum()

np.int64(13)

In [590]:
# Winners with more competitiveness for both 2021 & 2026

In [591]:
(margin_2021["winner_share%"] < 35).sum()

np.int64(2)

In [592]:
(margin_2026["winner_share%"] < 35).sum()

np.int64(64)

In [544]:
#creating competitive categories

def margin_bucket(x):
    if x < 5:
        return "Very Competitive (<5%)"
    elif x < 10:
        return "Competitive (5-10%)"
    elif x < 20:
        return "Moderate (10-20%)"
    else:
        return "Safe (>20%)"

In [593]:
margin_2026["competitive_type"]=(margin_2026["margin%"].apply(margin_bucket))

In [524]:
margin_2021["competitive_type"]=(margin_2021["vote_margin%"].apply(margin_bucket))

In [594]:
margin_2026["competitive_type"].value_counts()

competitive_type
Very Competitive (<5%)    103
Competitive (5-10%)        65
Moderate (10-20%)          49
Safe (>20%)                17
Name: count, dtype: int64

In [595]:
## To find the most competitive constituencies

margin_2026.sort_values(
    "margin%"
).head(10)

#This will give top 10 closest contest in Tamil Nadu

,ac_number,constituency,party_winner,votes_winner,party_runner,votes_runner,reserved,region,margin_votes,total_votes,margin%,winner_share%,competitive_type
233,185,Tirupattur,TVK,83375,DMK,83374,GEN,South,1,215710,0.000463,38.650000,Very Competitive (<5%)
221,54,Veppanahalli,DMK,74691,AIADMK,74553,GEN,North,138,222608,0.061992,33.552703,Very Competitive (<5%)
63,229,Kanniyakumari,AIADMK,75045,DMK,74831,GEN,South,214,239150,0.089484,31.379887,Very Competitive (<5%)
137,66,Polur,TVK,67961,DMDK,67734,GEN,North,227,211743,0.107205,32.095984,Very Competitive (<5%)
196,76,Tirukkoyilur,AIADMK,73033,TVK,72748,GEN,Central,285,217318,0.131144,33.606512,Very Competitive (<5%)
128,95,Paramathi-Velur,AIADMK,61349,DMK,61041,GEN,Central,308,188845,0.163097,32.486431,Very Competitive (<5%)
78,137,Kulithalai,DMK,68138,TVK,67559,GEN,Kongu,579,209091,0.276913,32.587725,Very Competitive (<5%)
60,80,Kallakurichi (SC),TVK,81132,AIADMK,80334,SC,Central,798,249626,0.319678,32.501422,Very Competitive (<5%)
80,171,Kumbakonam,TVK,78650,DMK,77971,GEN,Delta,679,208288,0.325991,37.760217,Very Competitive (<5%)
120,127,Palani,AIADMK,66986,TVK,66293,GEN,South,693,208619,0.332185,32.109252,Very Competitive (<5%)


In [596]:
## Most dominant victories

margin_2026.sort_values("margin%", ascending=False).head(10)

#This will give top 10 dominant victories

,ac_number,constituency,party_winner,votes_winner,party_runner,votes_runner,reserved,region,margin_votes,total_votes,margin%,winner_share%,competitive_type
42,86,Edappadi,AIADMK,148933,IND,50823,GEN,Central,98110,258263,37.988407,57.667184,Safe (>20%)
152,88,Salem (West),TVK,120407,PMK,45540,GEN,Central,74867,234407,31.938893,51.366640,Safe (>20%)
227,179,Viralimalai,AIADMK,105773,TVK,43700,GEN,Delta,62073,203921,30.439729,51.869597,Safe (>20%)
41,11,Dr.Radhakrishnan Nagar,TVK,97800,DMK,48132,GEN,Chennai Metro,49668,181206,27.409688,53.971723,Safe (>20%)
198,113,Tiruppur (North),TVK,131401,AIADMK,61409,GEN,Kongu,69992,263457,26.566764,49.875691,Safe (>20%)
132,12,Perambur,TVK,120365,DMK,66650,GEN,Chennai Metro,53715,204384,26.281411,58.891596,Safe (>20%)
85,9,Madavaram,TVK,190462,DMK,95477,GEN,Chennai Metro,94985,361998,26.239095,52.614103,Safe (>20%)
203,10,Tiruvottiyur,TVK,110067,CPI(M),56503,GEN,Chennai Metro,53564,207088,25.865333,53.149869,Safe (>20%)
138,2,Ponneri (SC),TVK,110439,INC,54671,SC,Chennai Metro,55768,226819,24.587006,48.690365,Safe (>20%)
232,50,Tirupattur,TVK,105098,DMK,56835,GEN,North,"48,263",202925,23.780000,51.790000,Safe (>20%)


In [597]:
## Region level competitiveness

margin_2026.groupby("region")["margin%"].mean().sort_values()

#This gives the average victory margin for each region. 
#So, we can see the chennai metro has highest victory margin and it is considered as very less competitve

region
South             5.820481
Kongu             6.375793
Delta             6.589305
Central           6.609037
North             6.675527
Chennai Metro    15.997587
Name: margin%, dtype: float64

##### - Chennai Metro exhibited significantly larger victory margins than any other region.
- Chennai Metro experienced extensive winner changes, yet the winning candidates often secured relatively comfortable victory margins.
- Outside Chennai Metro, victory margins were remarkably consistent across Tamil Nadu's regions, generally ranging between 5% and 7%.

### Insights

- Average victory margins narrowed from 11% in 2021 to 7% in 2026, indicating that constituencies were generally decided by closer contests in 2026.
- The number of constituencies where the winner secured an outright majority (>50% of votes) declined sharply from 70 to 13.
- Constituencies won with less than 35% vote share increased from 2 to 64, indicating a rise in fragmented electoral outcomes.

#### To tell this in the presentation, we can say that --

- The number of constituencies where the winner secured more than 50% of the total votes declined sharply. In 2021, there were 70 constituencies where the candidate secure more than 50% of the votes. And in 2026, that constituency decreased to 13 and it is 81.42%.
- Whereas the number of constituencies where the winner secured less than 35% of the total votes was increeased by 32 times in 2026 as compared to 2021. In 2021, there were only 2 constituency where the winning candidate had less than 35% vote share. And in 2026, there were 64 constituency were winning candidate had less than 35% of total vote share. This indicates a fragmented electoral outcome.
- Fragmented electoral outcome simply means votes are spread across many parties/candidates instead being concentrated on one or two.
- So, here we can say that 2026 elections where less fragmented as compared to 2021 elections.


- The 2026 election was characterized by narrower victory margins, fewer majority-winning constituencies, and a larger number of fragmented contests compared with 2021.

In [598]:
## Competitiveness of the victory margin of a winner candidate from the runner candidate.

margin_2026["competitive_type"].value_counts()

competitive_type
Very Competitive (<5%)    103
Competitive (5-10%)        65
Moderate (10-20%)          49
Safe (>20%)                17
Name: count, dtype: int64

In [599]:
margin_2021["competitive_type"].value_counts()

competitive_type
Moderate (10-20%)         75
Very Competitive (<5%)    67
Competitive (5-10%)       55
Safe (>20%)               37
Name: count, dtype: int64

### Insights 

- There are 103 constiuency which experienced high competitiveness and 17 constituency where the party won with high margin of victory in 2026.
- In 2021, the elections where moderative, the parties won with comparatebly very high margins as compared to 2026.